torch/distributed/pipelining/PipelineSchedule.py

```diff
@@ -694,161 +830,86 @@ def _step_microbatches(
+            if action is not None:
+                computation_type, mb_index, stage_index = action
+                if computation_type == _ComputationType.FORWARD:
+                    # perform forward computation
+                    stage = stage_index_to_stage[stage_index]
+                    output = stage.forward_one_chunk(
+                        arg_mbs[mb_index], kwarg_mbs[mb_index]
+                    )
+                    self._maybe_compute_loss(stage, output, target_mbs, mb_index)
+                    ops.extend(stage.get_fwd_send_ops())
+                elif computation_type == _ComputationType.BACKWARD:
+                    # perform backward computation
+                    stage = stage_index_to_stage[stage_index]
+                    loss = self._maybe_get_loss(stage, mb_index)
+                    stage.backward_one_chunk(loss=loss)
+                    ops.extend(stage.get_bwd_send_ops())
+                else:
+                    raise ValueError(f"Unknown computation type {computation_type}")
+
+            # Look at the neighboring ranks for this current timestep and determine whether
+            # this current rank needs to do any recv communication
+            prev_rank_action = None
+            if time_step < len(prev_rank_ops):
+                prev_rank_action = prev_rank_ops[time_step]
+            if prev_rank_action is not None:
+                computation_type, mb_index, stage_index = prev_rank_action
+                # Only handle sends for the forward from a previous rank
+                if computation_type == _ComputationType.FORWARD:
+                    # If not the last stage, then receive fwd activations
+                    if stage_index != self._num_stages - 1:
+                        # TODO: We are assuming that stage will always receive from stage-1
+                        # however that is not necessarily true of get_fwd_recv_ops
+                        stage = stage_index_to_stage[stage_index + 1]
+                        ops.extend(stage.get_fwd_recv_ops())
+                elif computation_type == _ComputationType.BACKWARD:
+                    # Previous rank doing backward has no influence for the current rank forward recv
+                    pass
+                else:
+                    raise ValueError(f"Unknown computation type {computation_type}")
+
+            next_rank_action = None
+            if time_step < len(next_rank_ops):
+                next_rank_action = next_rank_ops[time_step]
+            if next_rank_action is not None:
+                computation_type, mb_index, stage_index = next_rank_action
+                # Only handle receives for the backwards from a next rank
+                if computation_type == _ComputationType.FORWARD:
+                    # Next rank doing forward has no influence for the current rank backward recv
+                    pass
+                elif computation_type == _ComputationType.BACKWARD:
```

Comment:
_configure_data_parallel_mode is missing and needs to be added back

Reply:
we should add a test to our schedule tests that checks if configure_data_parallel is called properly. (another PR)

Reply:
Thanks, good catch. I added it back in

Reply:
Will follow up with a test for it too